### ISSUE TITLE
sending custom payload returns 422 error

### ISSUE THREAD

**Issue by ikbenben** [benedikt/layer-ruby#9]
Hi There,

I haven't determined if this a layer-ruby issue or an issue with the platform API so thought I would start here. Yea re trying to send a custom payload to Layer as follows:

      payload = {
        sender: { user_id: sender.id.to_s },
        parts: [
          { mime_type: "metadata", body: { message_type: "simple-text" } },
          { mime_type: "text/plain", body: message }
        ]
      }

      conversation.messages.create(payload)

when I run this code, I get the following:

    RestClient::UnprocessableEntity: 422 Unprocessable Entity  
	   from /usr/local/bundle/gems/rest-client-1.8.0/lib/restclient/abstract_response.rb:74:in `return!'  
	   from /usr/local/bundle/gems/rest-client-1.8.0/lib/restclient/request.rb:495:in `process_result'  
	   from /usr/local/bundle/gems/rest-client-1.8.0/lib/restclient/request.rb:421:in `block in transmit'  

I'm not sure if the gem is constructing the payload wrong when I include the custom part. It works if I just send a message with a single part of mime_type = text/plain.

thanks

**Comment by benedikt** [comment]
Hi @ikbenben!

This isn't an issue with the gem itself, but with how you use the API. As far as I know, you have to base64 encode the body if it's not a string already.

You might want to upgrade layer-ruby to the most recent revision on GitHub to get improved error messages, that are more helpful when debugging this.

```ruby
gem 'layer-ruby', github: 'benedikt/layer-ruby'
```

Hope this helps!

Cheers,

Benedikt

**Comment by ikbenben** [comment]
Hi Benedikt,

thanks for the reply. the message in the text/plain is just a string. are you suggesting that the 'body' attribute for our mime_type: "metadata" needs to be base64 encoded?

**Comment by benedikt** [comment]
Yes, you have to encode the metadata, as `body` must be a String. It probably doesn't have to be base64, but at least encoded with `.to_json`:

```ruby

payload = {
  sender: { user_id: sender.id.to_s },
  parts: [
    { mime_type: "metadata", body: { message_type: "simple-text" }.to_json },
    { mime_type: "text/plain", body: message }
  ]
}

conversation.messages.create(payload)
```

**Comment by ikbenben** [comment]
Hi Benedikt,

One follow up question. Sorry :) 

I was starting with the custom mime type first but I'm also having trouble sending a part with a mime_type of image/png. Can you see anything specifically wrong with this?

     payload = {
        sender: { user_id: sender.id.to_s },
        parts: [
          { mime_type: "text/plain", body: message },
          { mime_type: "image/png", body: get_badge_image_as_base64(badge), encoding: "base64" }
        ]
      }

    def get_badge_image_as_base64(badge)

      image = Rails.application.assets["motivations_#{badge}@3x.png"]
      image_as_string = open(image.pathname) { |d| d.read }
      base64 = ActiveSupport::Base64.encode64(image_as_string)

      base64

    end

**Comment by ikbenben** [comment]
I determined that it was the encoding property that was being sent but the Platform API documentation says to use that property if sending base64 encoded string and their samples show using it...any thoughts?

**Comment by ikbenben** [comment]
Hi Benedikt. Do you have any thoughts here?

**Comment by ikbenben** [comment]
Never mind, I have recreated the request using curl, eliminating your gem and it is not related to your gem. Thanks again

**Comment by benedikt** [comment]
Hi @ikbenben,

sorry for not getting back to you yesterday. I'm glad you figured out the issue. 

I just released layer-ruby v0.5.0 which provides improved error messages. They now show the error response provided by Layer, which greatly helps when something goes wrong.

Bests,

Benedikt
